test/extensions/quic_listeners/quiche/envoy_quic_alarm_test.cc

```diff
@@ -0,0 +1,148 @@
+#include "common/event/libevent_scheduler.h"
+
+#include "extensions/quic_listeners/quiche/envoy_quic_alarm.h"
+#include "extensions/quic_listeners/quiche/envoy_quic_alarm_factory.h"
+#include "extensions/quic_listeners/quiche/platform/envoy_quic_clock.h"
+
+#include "test/test_common/simulated_time_system.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using Envoy::Event::Dispatcher;
+using quic::QuicTime;
+
+namespace Envoy {
+namespace Quic {
+
+class TestDelegate : public quic::QuicAlarm::Delegate {
+public:
+  TestDelegate() : fired_(false) {}
+
+  void OnAlarm() override { fired_ = true; }
+  bool fired() const { return fired_; }
+
+private:
+  bool fired_;
+};
+
+class EnvoyQuicAlarmTest : public ::testing::Test {
```

Comment:
Some alarm library "eagerly fires" when a alarm deadline is <= now. It seems like this implementation doesn't do that, can you add a test to ensure that behavior?

Reply:
What does "eagerly fire" mean? Current implementation do schedule the alarm in same event loop. Added test FireAlarmWithPastDeadline

Reply:
Thanks for adding the test:)

By "eagerly fire" I mean if you schedule an alarm to fire "in the past", the schedule api will call the Fire() callback before it returns, for such alarms we need to make sure the callback is safe to run in both the calling thread and the event loop.